src/internal/small_map.rs

```diff
@@ -0,0 +1,193 @@
+            Self::Two(_) => None,
+            Self::Flexible(data) => data.get_mut(key),
+        }
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        let s = std::mem::replace(self, Self::Flexible(Default::default()));
+        let out;
+        *self = match s {
+            Self::Empty => {
+                out = None;
+                Self::Empty
+            }
+            Self::One([(k, v)]) => {
+                if key == &k {
+                    out = Some(v);
+                    Self::Flexible(Default::default())
+                } else {
+                    out = None;
+                    Self::One([(k, v)])
+                }
+            }
+            Self::Two([(k1, v1), (k2, v2)]) => {
+                if key == &k1 {
+                    out = Some(v1);
+                    Self::One([(k2, v2)])
+                } else if key == &k2 {
+                    out = Some(v2);
+                    Self::One([(k1, v1)])
+                } else {
+                    out = None;
+                    Self::Two([(k1, v1), (k2, v2)])
+                }
+            }
+            Self::Flexible(mut data) => {
+                out = data.remove(key);
+                Self::Flexible(data)
+            }
+        };
+        out
+    }
+
+    pub fn insert(&mut self, key: K, value: V) {
+        let s = std::mem::replace(self, Self::Flexible(Default::default()));
+        *self = match s {
+            Self::Empty => Self::One([(key, value)]),
+            Self::One([(k, v)]) => {
+                if key == k {
+                    Self::One([(k, value)])
+                } else {
+                    Self::Two([(k, v), (key, value)])
+                }
+            }
+            Self::Two([(k1, v1), (k2, v2)]) => {
+                if key == k1 {
+                    Self::Two([(k1, value), (k2, v2)])
+                } else if key == k2 {
+                    Self::Two([(k1, v1), (k2, value)])
+                } else {
+                    let mut data: Map<K, V> = Map::with_capacity_and_hasher(3, Default::default());
+                    data.insert(key, value);
+                    data.insert(k1, v1);
+                    data.insert(k2, v2);
+                    Self::Flexible(data)
+                }
+            }
+            Self::Flexible(mut data) => {
+                data.insert(key, value);
+                Self::Flexible(data)
+            }
+        };
+    }
+}
+
+impl<K: Clone + PartialEq + Eq + Hash, V: Clone> SmallMap<K, V> {
+    /// Merge two hash maps.
+    ///
+    /// When a key is common to both,
+    /// apply the provided function to both values.
+    /// If the result is None, remove that key from the merged map,
+    /// otherwise add the content of the Some(_).
+    pub fn merge<'a>(
+        &'a mut self,
+        map_2: impl Iterator<Item = (&'a K, &'a V)>,
+        f: impl Fn(&V, &V) -> Option<V>,
+    ) {
+        for (key, val_2) in map_2 {
```

Comment:
Depending on the order of the keys, considering two `SmallMap` of 2 elements, there is a 50% chance that it grows to 3 elements before removing a key back to 2. If we could find a benchmark with high conflict rate, it may be interesting to check if increasing the small map enum to 3 or 4 elements might be worth it.